### PR TITLE
[AKS] add support for undrainable_node_behavior

### PR DESCRIPTION
**Pull request by zjpjack-github**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### Related command

az aks nodepool add/update/upgrade --undrainable-node-behavior

### General Guidelines

- [ ] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [ ] Have you run `python scripts/ci/test_index.py -q` locally? (`pip install wheel==0.30.0` required)
- [ ] My extension version conforms to the [Extension version schema](https://github.com/Azure/azure-cli/blob/release/doc/extensions/versioning_guidelines.md)

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
You only need to update the version information in file setup.py and historical information in file HISTORY.rst in your PR but do not modify `src/index.json`.

### CODE REVIEW

**Review by FumingZhang — COMMENTED** [review]
lgtm, please resolve merge conflict

Queued [live test](https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=103002390&view=results) to validate the change.

- test_aks_nodepool_undrainable_node_behavior

File: src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py [hunk 1]
```diff
@@ -3108,6 +3108,68 @@ def test_aks_nodepool_node_soak_duration(
...
+        self.cmd(
+            create_cmd,
+            checks=[
+                self.check("provisioningState", "Succeeded"),
+            ],
+        )
+
+        add_nodepool_cmd = (
+            "aks nodepool add -g {resource_group} --cluster-name {name} -n {nodepool_name} "
+            "--mode user --undrainable_node_behavior Cordon"
```
Comment (FumingZhang) [inline src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py:3143]:
```suggestion
            "--mode user --undrainable-node-behavior Cordon"
```

File: src/aks-preview/HISTORY.rst [hunk 2]
```diff
@@ -13,6 +13,10 @@ Pending
 +++++++
 * Vendor new SDK and bump API version to 2024-07-02-preview.
```
Comment (FumingZhang) [inline src/aks-preview/HISTORY.rst:14]: Please include this line to the new version you added.

**Review by FumingZhang — COMMENTED** [review]
Linter failed with following error, you could bypass them by updating `src/aks-preview/linter_exclusions.yml`
```
-  FAIL - HIGH severity: option_length_too_long
    Parameter: aks nodepool add, `undrainable_node_behavior` - The lengths of all options ['--undrainable-node-behavior'] are longer than threshold 22. Argument undrainable_node_behavior must have a short abbreviation.
    Parameter: aks nodepool update, `undrainable_node_behavior` - The lengths of all options ['--undrainable-node-behavior'] are longer than threshold 22. Argument undrainable_node_behavior must have a short abbreviation.
    Parameter: aks nodepool upgrade, `undrainable_node_behavior` - The lengths of all options ['--undrainable-node-behavior'] are longer than threshold 22. Argument undrainable_node_behavior must have a short abbreviation.
```

Please commit the recording file according to your newly added test case to `src/aks-preview/azext_aks_preview/tests/latest/recordings/`
image

File: src/aks-preview/HISTORY.rst [hunk 3]
```diff
@@ -12,6 +12,7 @@ To release a new version, please select a new version number (usually plus 1 to
 9.0.0b2
 +++++++
 * Add block to supported outbound type
+* Add `--undrainable-node-behavior` to the `az aks nodepool add/update/upgrade` commands.
```
Comment (FumingZhang) [inline src/aks-preview/HISTORY.rst:15]: 9.0.0b2 has been released, please choose a new version

File: src/aks-preview/azext_aks_preview/_params.py [hunk 4]
```diff
@@ -1529,6 +1529,7 @@ def load_arguments(self, _):
         c.argument("max_surge", validator=validate_max_surge)
         c.argument("drain_timeout", type=int)
         c.argument("node_soak_duration", type=int)
+        c.argument("undrainable_node_behavior", type=str)
```
Comment (FumingZhang) [inline src/aks-preview/azext_aks_preview/_params.py:1532]: The default type is str, could omit `type=str`.

File: src/aks-preview/azext_aks_preview/_params.py [hunk 5]
```diff
@@ -1641,6 +1642,7 @@ def load_arguments(self, _):
...
```
Comment (FumingZhang) [inline src/aks-preview/azext_aks_preview/_params.py:1645]: same as above

File: src/aks-preview/azext_aks_preview/_params.py [hunk 6]
```diff
@@ -1710,6 +1712,7 @@ def load_arguments(self, _):
...
```
Comment (FumingZhang) [inline src/aks-preview/azext_aks_preview/_params.py:1715]: same as above

File: src/aks-preview/HISTORY.rst (hunk 3 above)
Comment (FumingZhang) [reply src/aks-preview/HISTORY.rst:15]: and update `VERSION` in setup.py

**Review by FumingZhang — APPROVED** [review]
LGTM

File: src/aks-preview/azext_aks_preview/_help.py [hunk 7]
```diff
@@ -1888,6 +1888,9 @@
         - name: --vm-sizes
           type: string
           short-summary: Comma-separated list of sizes. Must use VirtualMachines agent pool type.
+        - name: --undrainable-node-behavior
+          type: string
+          short-summary: Defines the behavior for undrainable nodes during upgrade. The value should be "Cordon" or "Schedule". The default value is "Schedule".
```
Comment (yanzhudd) [inline src/aks-preview/azext_aks_preview/_help.py:1893]: please be consistant with other parameters and use first-person verb.
It is better to define this parameter with `arg_type=get_enum_type()` if it is enumeable.
```suggestion
          short-summary: Define the behavior for undrainable nodes during upgrade. The default value is "Schedule".
```

File: src/aks-preview/azext_aks_preview/_help.py [hunk 8]
```diff
@@ -1952,6 +1955,9 @@
         - name: --if-none-match
           type: string
           short-summary: Set to '*' to allow a new node pool to be created, but to prevent updating an existing node pool. Other values will be ignored.
+        - name: --undrainable-node-behavior
+          type: string
+          short-summary: Defines the behavior for undrainable nodes during upgrade. The value should be "Cordon" or "Schedule". The default value is "Schedule".
```
Comment (yanzhudd) [inline src/aks-preview/azext_aks_preview/_help.py:1960]: the same as above

File: src/aks-preview/azext_aks_preview/_help.py [hunk 9]
```diff
@@ -2043,6 +2049,9 @@
         - name: --disable-fips-image
           type: bool
           short-summary: Switch to use non-FIPS-enabled OS on agent nodes.
+        - name: --undrainable-node-behavior
+          type: string
+          short-summary: Defines the behavior for undrainable nodes during upgrade. The value should be "Cordon" or "Schedule". The default value is "Schedule".
```
Comment (yanzhudd) [inline src/aks-preview/azext_aks_preview/_help.py:2054]: the same as above

**Review by yanzhudd — APPROVED** [review]

**Pull request merged**